how-it-works/7_smart_contracts/01_canister-lifecycle.subpage.md

```diff
@@ -0,0 +1,40 @@
+The code of a canister consists of a Wasm module. The state consists of the usual heap of the Wasm module, together with stable memory, a special type of memory which plays an important role in the life-cycle of a canister.
+
+## How do canisters work
+
+Canisters behave much like actors from the actor-based concurrency model.  Their code is single threaded and is executed in complete isolation of other canisters. Canisters communicate with one another via asynchronous messaging.  When processing a message, a canister can make changes to its state, trigger the sending of messages to other canisters, or even create other canisters. Unlike in the traditional actor model, communication is bidirectional. Canister messages are either requests or replies. For each request sent, a canisters records with the IC a callback to be invoked when the callee sends back a response.  If the IC determines that there is no way for the callee to respond then the system will produce a response.
+Another novel aspect of the canister based model is the interplay between message processing and canister trapping.  While processing a request a canister may send requests to other canisters and wait for (some of) the replies, before producing a reply to the original request.
+If a canister traps, its state is rolled back to the point right after it made the last outgoing call.
+
+
+## Resource charging
+As they execute, canisters use resources in the form of memory, computation and network bandwidth.  On the IC all of these are paid for using a unit called cycles. To this end, each canister has a local cycles account from which the system deducts cycles as execution proceeds. Charging for memory usage is straightforward.  The system keeps track of the memory used by the canister and regularly charges the canister’s balance.  For efficiency, this charging happens at regular intervals but not every round. In contrast, charging for computation happens when computation performs. To this end, the canisters are instrumented with code that allows the IC to count the number of instructions executed. Every round, there is an upper bound on the number of instructions which execution can perform every round.  If this number of instructions is exceeded, then execution is paused and continued in a subsequent round. However,  cycles for the computation performed during any round are already charged at the end of that round. To prevent a buggy or malicious canister from completely taking over an execution core, the total number of rounds the execution of a canister can take is also bounded.
```

Comment:
```suggestion
As they execute, canisters use resources in the form of memory, computation and network bandwidth.  On the IC all of these are paid for using a unit called *cycles*. To this end, each canister has a local cycles account from which the system deducts cycles as execution proceeds. Charging for memory usage is straightforward.  The system keeps track of the memory used by the canister and regularly charges the canister’s balance.  For efficiency, this charging happens at regular intervals but not every round. In contrast, charging for computation happens when computation performs. To this end, the canisters are instrumented with code that allows the IC to count the number of instructions executed. Every round, there is an upper bound on the number of instructions which execution can perform every round.  If this number of instructions is exceeded, then execution is paused and continued in a subsequent round. However,  cycles for the computation performed during any round are already charged at the end of that round. To prevent a buggy or malicious canister from completely taking over an execution core, the total number of rounds the execution of a canister can take is also bounded.
```